README.md
**Current Version** <br />
**Script:** `2.0` (May 21, 2017) <br />
**Service:** `2.0` (May 21, 2017) <br />

**Note: The 2.0+ Script/Service files wont work with versions older than 2.0 due to big changes**

# Description
Easy to use Script that can set Windows 10 services based on Black Viper's Service Configurations. <br />

Black Viper's Service Configurations from http://www.blackviper.com/

**Note 1: Script is ment for Windows 10 Home x64 / Windows 10 Pro x64.. With Creator's Update Installed**  <br />
**Note 2: If you are using a non english version of Windows and get a Build Check Error, please let me know**  <br />

PS. Don't forget to check out my other Repo https://github.com/madbomb122/Win10Script  <br />

# Requirements (Make sure to LOOK at this)

|            | Recommended Requirements         | AT YOURN OWN RISK* (See Note Bellow) |
| :--------- | :--------------------------------| :--------------------------- |
|**OS**      | Windows 10 x64                   | Windows 10 x32               |
|**Edition** | Pro or Home                      | All Others                   |
|**Build**   | Creator's Update                 | Pre-Creator's Update         |

<br />

**_Need Files_** <br />
`BlackViper-Win10.ps1` (Script) <br />
`BlackViper.csv` (Service Configurations) <br />

**Recommended Files** <br />
`_Win10-BlackViper.bat` (To run script easier) <br /> 
`README.md` (This Readme) <br />

**Note:**
**AT YOUR OWN RISK** <br />
**1. Run the script on x32 w/o changing settings (But shows a warning)** <br />
**2. Skip the check for** <br />
**---A. Home/Pro** *(`$Script:Edition_Check` variable in script or use `-sec` switch)* <br />
**---B. Creator's Update** *(`$Script:Build_Check` variable in script or use `-sbc` switch)* <br />

# How to Use/Run
Download/Save the following files <br />
Script File: `BlackViper-Win10.ps1` -Need <br />
Service Config File: `BlackViper.csv` -Need <br />
Bat File: `_Win10-BlackViper.bat` -Recommended <br />
  **Note: DO NOT RENAME THE FILES**<br />
Next follow the **Basic Usage** or **Advanced Usage**

# Basic_Usage
Run the Script by bat file `_Win10-BlackViper.bat` (Recommended) <br />
`powershell.exe -NoProfile -ExecutionPolicy Bypass -File BlackViper-Win10.ps1` <br />
Select desired Services Configuration <br />
*1. Default <br />
2. Safe (Recommended Option) <br />
3. Tweaked (Not supported for laptop ATM)<br />*

# Advanced_Usage
Use one of the following Methods you can 
1. Run script or bat file with one (or more) of the switches bellow
2. Edit the script (bottom of file) to change the values
3. Edit the bat file (top of file) to change the values to add the switch


|   Switch  | Description                                                                    | Notes                            |
| :-------- | :------------------------------------------------------------------------------| :------------------------------- |
| -atos     | Accepts the ToS                                                                |                                  |
| -auto     | Runs the script to be Automated.. Closes on User input, Errors, End of Script) | Implies `-atos`                    |
| -default  | Runs the script with Services to Default Configuration                         | Alt `-set default` or `-set 1` |
| -safe     | Runs the script with Services to Black Viper's Safe Configuration              | Alt `-set safe` or `-set 2`    |
| -tweaked  | Runs the script with Services to Black Viper's Tweaked Configuration           | Alt `-set tweaked` or `-set 3` |
| -all      | Every windows services will change                                             |   |
| -min      | Just the services different from the default to safe/tweaked list              |   |
| -sec      | Skips Edition Check (Home/Pro)                                                 | **USE AT YOUR OWN RISK**             |
| -sbc      | Skips Build Check (Creator's Update)                                           | **USE AT YOUR OWN RISK**             |
| -sic      | Skips Internet Check (If checking for update)                                  | Tests by pinging github.com      |
| -usc      | Checks for Update to Script file before running                                | Auto downloads and runs if found |
| -use      | Checks for Update to Service file before running                               | Auto downloads and uses if found |
| -diag     | Shows some diagnostic information on error messages                            | **Stops automation** |
| -log      | Makes a log file (Logs Notices, Errors, & Services changed)                    | Log file `Script.log` (default) |
| -baf      | File of all the services before and after the script                           | `Services-Before.log` and `Services-After.log` |

Examples: <br />
`powershell.exe -NoProfile -ExecutionPolicy Bypass -File BlackViper-Win10.ps1 -Set Default` <br />
`powershell.exe -NoProfile -ExecutionPolicy Bypass -File BlackViper-Win10.ps1 -atos -use -set 2` <br />
`powershell.exe -NoProfile -ExecutionPolicy Bypass -File BlackViper-Win10.ps1 -auto -use -tweaked -sec` <br />
`_Win10-BlackViper.bat -Set Default` <br />
`_Win10-BlackViper.bat -atos -use -set 2` <br />
`_Win10-BlackViper.bat -auto -use -tweaked -sec` <br />
******

# FAQ

**Q:** The script file looks all messy in notepad, How do i view it? <br />
**A:** Try using wordpad or what I recommend, Notepad++ https://notepad-plus-plus.org/

**Q:** Do you accept any donations? <br />
**A:** I'd rather you donate to Black Viper... but if you would like to donate to me I accept amazon giftcards (email it to me). <br />
**NOTE:** http://www.blackviper.com/support-bv/

**Q:** How can I contact you? <br />
**A:** You can email me @ madbomb122@gmail.com, if it's an issue please post under issue

**Q:** The script wont run, can you help me? <br />
**A:** Yes, but first if you are using automation.. turn off automation and see if it gives and error that you can correct.

**Q:** The script window closes or gives an error saying script is blocked, what do i do? <br />
**A:** By default windows blocks ps1 scripts, you can use one of the following <br />
         1. Use the bat file to run the script (recommended) <br />
         2. On an admin powershell console `Set-ExecutionPolicy -Scope CurrentUser -ExecutionPolicy Unrestricted` <br />	 

**Q:** Who do I contact about the Service Configurations or an issue with the configuration? <br />
**A:** Any "technical" issues (or why something is set one way or another) can be directed to Black Viper himself.

**Q:** Why does you script not change the service *BLAH*? <br />
**A:** You didnt select the All option, it's not a default windows service, cant be changed, or some other good reason.

**Q:** I have an issue with the script, what do I do? <br />
**A:** Post it as an issue using github's issues tab up top.

**Q:** Can I run the script safely? <br />
**A:** The script itself is safe, but changes to services may cause problems for certin programs.

**Q:** Can I run the script repeatedly? <br />
**A:** Yes, with same or different settings.

**Q:** I've run the script and it did *BLAH*, can I undo it? <br />
**A:** Yes, run the script again and select again. <br />

**Q:** Can I use the script or modify it for my / my company's needs? <br />
**A:** Sure. Just don't forget to include copyright notice as per the license requirements, and leave any Copyright in script too.

**Q:** The script messed up my computer. <br />
**A:** The script is as is, any problems you have/had is your own problem.

**Q:** Are you going to add support for other editions of Windows 10 other than Pro or Home? <br />
**A:** Sorry, I only support the windows 10 configuration listed on Black Viper's website. <br />

**Q:** Are you going to add support for builds before the "Creator's Update"? <br />
**A:** Sorry, since I dont have the configuration from before the "Creator's Update" I can't add it. <br />
**Note:** If you have the configuration please contact me.

**Q:** Will you make a script for any windows before windows 10? <br />
**A:** No. <br />

**Q:** Can I download the csv file from Black Viper's website and use that? <br />
**A:** No, my file is not the same. <br />

**Q:** Can I add a service to be changed or stop one from changing? <br />
**A:** Yes, to add/remove edit the file `BlackViper.csv` <br />
---to remove a service remove the line or put something to change service name, other than symbols (# is fime) <br />
---to add put it in the proper format <br />
**Note:** Number meaning `0 -Not Installed/Skip`, `1 -Disable`, `2 -Manual`, `3 -Automatic`, `4 -Auto (Delayed)` <br />

**Q:** I have a suggestion for the script, how do i suggest it? <br />
**A:** Do a pull request with the change or submit it as an issue with the suggestion. <br />

**Q:** How long are you going to maintain the script? <br />
**A:** No Clue.

**Q:** When do you update the services file (BlackViper.csv)? <br />
**A:** When Black Viper tells me a change, I see an update on his site, or someone tells me there is an update.


Changelog.txt
-----------------------------------------------------------------------------------
1.6 (05-14-2017)

Fixed a bug with logging that caused it to not continue to log after script update

-----------------------------------------------------------------------------------
1.5 (05-12-2017)

Minor Fixes (didnt affect using script)
Added ability to have a log file (Logs notices, Errors, & Services changed)
Added Argument/Switch for log

 Switches       Description of Switch
  -log           Saves a log file (Script.log by default)

-----------------------------------------------------------------------------------
1.4 (05-10-2017)

Fixed error in script was causing safe and tweaked option to not change services

Thanks to Scott C for notifying me about the bug

-----------------------------------------------------------------------------------
1.3 (05-06-2017)

Fixed a messup on the new switch/arguments system

-----------------------------------------------------------------------------------
1.2 (05-05-2017)

Added/Changed Arguments/Switches to accept multiple (Old Switches Work too)

 Switches       Description of Switch
  -atos          (Accepts ToS)
  -auto          (Runs the script to be Automated.. Closes on User input, Errors, End of Script)
  -default       (Runs the script with Services to Default Configuration)
  -safe          (Runs the script with Services to Black Viper's Safe Configuration)
  -tweaked       (Runs the script with Services to Black Viper's Tweaked Configuration)
  -sec           (Skips Edition Check)
  -sbc           (Skips Build Check)
  -sic           (Skips Internet Check)
  -usc           (Checks for Update to Script file before running)
  -use           (Checks for Update to Service file before running)
  -diag          (Shows diagnostic info)

-----------------------------------------------------------------------------------
1.1 (04-23-2017)

Various internal script changes
Added another check for 'BlackViper.csv', in a certin case script would have ran w/o it
Fixed download link to BlackViper.csv
Fixed an issue with script failing edition check on non english versions of windows
Fixed an issue with edition check showing blank error

Thanks to Mehanik511 posting an issue, both edition problems were fixed

-----------------------------------------------------------------------------------
1.0 (04-14-2017)

Updated script to check for laptop/desktop

-----------------------------------------------------------------------------------
0.9 (04-02-2017) -Final Beta/Test Version

Script checks for service file update by default
Some display changes
Changed the 'go to website' to open with default browser instead of IE
added bypass for internet check (incase you have google.com and yahoo.com blocked)

-----------------------------------------------------------------------------------
0.8 (03-29-2017)

Added Internet Check if script checks for update (pings google.com and yahoo.com)
Added Copyright Display Option
Goto Website Option
Changed TOS Display look

-----------------------------------------------------------------------------------
0.7 (03-22-2017)

Enabled Build Check (For creator's update)
Added Display of Script Version/Date
Added Display of Last update of Service File

Removed Display of Websites

Various script improvements

-----------------------------------------------------------------------------------
0.6 (03-19-2017)

Added variable with ability to check for update to Script File (Skips by default)
Note: If found will Auto download and runs that (file name will be "BlackViper-Win10-Ver.(version-here).ps1")

Added variable with ability to check for update to Service File (Skips by default)
Note: If found will Auto download and uses that for the configuration

-----------------------------------------------------------------------------------
0.5 (03-18-2017)

Added variable with ability to:
Skip Edition check (USE AT YOUR OWN RISK)
Skip Build check (USE AT YOUR OWN RISK)
Skip TOS (If you accept it)
have Script close at end (before it paused so you can see results)
show/hide display of Services change, already set, and not present

Added per Black Viper's Request
Build Check (For creator's update) -Disabled till creator's update is released

-----------------------------------------------------------------------------------
0.4 (03-14-2017)

Added Color to Script output
Fixed bug that showed "Service is already" if service doesnt exist

Changed script to load services from file 'BlackViper.csv' instead of from within script
^If it isnt with script it will give you an option to download the file

-----------------------------------------------------------------------------------
0.3 (03-13-2017)

Added per Black Viper's Request
Check for x64 bit (still runs on x32 but gives a warning) 

-----------------------------------------------------------------------------------
0.2 (03-04-2017)

Bug fix (script didnt get the right type for current service setting)

-----------------------------------------------------------------------------------
0.1 (02-26-2017)

Added ability to use switches to bypass menu to run one of the choices

Choices are:
-set 1
-set default

-set 2
-set safe

-set 3
-set tweaked

-----------------------------------------------------------------------------------
0.0 (02-22-2017)

Initial Start of Script


Old/Win10-BlackViper.ps1
##########
# Win10 Black Viper Service Configuration Script
# 
# Black Viper's Service Configurations From
# Website: http://www.blackviper.com/service-configurations/black-vipers-windows-10-service-configurations/
#
# Script + Menu By
# Author: Madbomb122
# Website: https://github.com/madbomb122/
# Version: 0.3, 03-14-2017
#
# Release Type: Beta
##########

<#
    Copyright (c) 2017 Madbomb122 - Black Viper Service Script
    Copyright (c) 1999-2017 by Charles "Black Viper" Sparks - Actual settings for the Services
    
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#>

Param([alias("Set")] [string] $SettingImp)

$ErrorActionPreference= 'silentlycontinue'
$RelType = "Beta"
#$RelType = "Testing"
#$RelType = "Stable"

# Ask for elevated permissions if required
If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]"Administrator")) {
    Start-Process powershell.exe "-NoProfile -ExecutionPolicy Bypass -File `"$PSCommandPath`" $args" -Verb RunAs
    Exit
}

Function TOSDisplay {
    Write-Host "                 Terms of Use                  " -ForegroundColor Green -BackgroundColor Black
    Write-Host "                                               " -ForegroundColor Black -BackgroundColor White
    Write-Host "This program comes with ABSOLUTELY NO WARRANTY." -ForegroundColor Black -BackgroundColor White
    Write-Host "This is free software, and you are welcome to  " -ForegroundColor Black -BackgroundColor White
    Write-Host "redistribute it under certain conditions.      " -ForegroundColor Black -BackgroundColor White
    Write-Host "                                               " -ForegroundColor Black -BackgroundColor White
    Write-Host "Read License file for full Terms.              " -ForegroundColor Black -BackgroundColor White
    Write-Host "                                               " -ForegroundColor Black -BackgroundColor White
    If ($OSType -ne 64){
        Write-Host "                                               " -ForegroundColor Red -BackgroundColor Black
        Write-Host "                 WARNING!!!                    " -ForegroundColor Yellow -BackgroundColor Black
        Write-Host "     These settings are ment for x64 Bit.      " -ForegroundColor Red -BackgroundColor Black
        Write-Host "             Use AT YOUR OWN RISK.             " -ForegroundColor Red -BackgroundColor Black
        Write-Host "                                               " -ForegroundColor Red -BackgroundColor Black
    }
    If ($RelType -ne "Stable"){
        Write-Host "                                               " -ForegroundColor Red -BackgroundColor Black
        Write-Host "                 Caution!!!                    " -ForegroundColor Yellow -BackgroundColor Black
        Write-Host " Service Configuration are based on Creator's  " -ForegroundColor Red -BackgroundColor Black
        Write-Host " Update and is still being changed.            " -ForegroundColor Red -BackgroundColor Black
        Write-Host " Use AT YOUR OWN RISK.                         " -ForegroundColor Red -BackgroundColor Black
    }
}

function TOS {
    $TOS = 'X'
    while($TOS -ne "Out"){
        Clear-Host
        TOSDisplay
        If($Invalid -eq 1){
            Write-host ""
            Write-host "Invalid Input" -ForegroundColor Red -BackgroundColor Black -NoNewline
            $Invalid = 0
        }
        $TOS = Read-Host "`nAccept? (Y)es/(N)o"
        switch ($TOS.ToLower()) {
            n {Exit}
            no {Exit}
            y {Black_Viper_Input}
            yes {Black_Viper_Input}
            default {$Invalid = 1}
        }
    }
    Return
}

Function DisplayOutMenu([String]$TxtToDisplay,[int]$TxtColor,[int]$BGColor,[int]$NewLine){
    If ($NewLine -eq 0){
        Write-Host -NoNewline $TxtToDisplay -ForegroundColor $colors[$TxtColor] -BackgroundColor $colors[$BGColor]
    } Else {
        Write-Host $TxtToDisplay -ForegroundColor $colors[$TxtColor] -BackgroundColor $colors[$BGColor]
    }
}

function ChoicesDisplay ([Array]$ChToDisplay) {
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
    DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[0] 11 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
    DisplayOutMenu "|                                                   |" 14 0 1
    If ($OSType -ne 64){
        DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[2] 2 0 0 ;DisplayOutMenu "|" 14 0 1
        DisplayOutMenu "|                                                   |" 14 0 1
        DisplayOutMenu "|" 14 0 0 ;DisplayOutMenu "  Settings are ment for x64. Use AT YOUR OWN RISK. " 13 0 0 ;DisplayOutMenu "|" 14 0 1
    } Else {
        DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[1] 2 0 0 ;DisplayOutMenu "|" 14 0 1
        DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[2] 2 0 0 ;DisplayOutMenu "|" 14 0 1
	}
    DisplayOutMenu "|                                                   |" 14 0 1
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
    DisplayOutMenu "|                                                   |" 14 0 1
    for ($i=3; $i -le 5; $i++) {
        DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[$i] 14 0 0 ;DisplayOutMenu "|" 14 0 1
    }
    DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[6] 13 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|                                                   |" 14 0 1
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
    DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[7] 15 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|  " 14 0 0 ;DisplayOutMenu $ChToDisplay[8] 15 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
    DisplayOutMenu "|" 14 0 0 ;DisplayOutMenu "          https://github.com/madbomb122/           " 15 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|" 14 0 0 ;DisplayOutMenu "            http://www.blackviper.com/             " 15 0 0 ;DisplayOutMenu "|" 14 0 1
    DisplayOutMenu "|---------------------------------------------------|" 14 0 1
}

function OpenWebsite ([String]$Website){
    $IE=new-object -com internetexplorer.application
    $IE.navigate2($Website)
    $IE.visible=$true
}

function Black_Viper_Input {
    $Black_Viper_Input = 'X'
    while($Black_Viper_Input -ne "Out"){
        Clear-Host
        ChoicesDisplay $BlackViperDisItems
        If($Invalid -eq 1){
            Write-host ""
            Write-host "Invalid Input" -ForegroundColor Red -BackgroundColor Black -NoNewline
            $Invalid = 0
        }
        $Black_Viper_Input = Read-Host "`nChoice"
        switch ($Black_Viper_Input.ToLower()) {
            1 {Black_Viper_Set 1; $Black_Viper_Input = "Out"}
            2 {Black_Viper_Set 2; $Black_Viper_Input = "Out"}
            3 {Black_Viper_Set 3; $Black_Viper_Input = "Out"}
            M {OpenWebsite "https://github.com/madbomb122/Win10Script/"}
            B {OpenWebsite "http://www.blackviper.com/"}
            Q {Exit}
            default {$Invalid = 1}
        }
    }
}

If ([System.Environment]::Is64BitProcess) {
    $OSType = 64
}

$BlackViperDisItems = @(
"       Black Viper's Service Configurations      ",
"                                                 ",
"Settings based on Black Viper's Configurations.  ",
'1. Default                                       ',
'2. Safe                                          ',
'3. Tweaked                                       ',
'Q. Quit (No changes)                             ',
"M. Go to Madbomb122's Github                     ",
"B. Go to Black Viper's Website                   ")

$colors = @(
    "black",        #0
    "blue",         #1
    "cyan",         #2
    "darkblue",     #3
    "darkcyan",     #4
    "darkgray",     #5
    "darkgreen",    #6
    "darkmagenta",  #7
    "darkred",      #8
    "darkyellow",   #9
    "gray",         #10
    "green",        #11
    "magenta",      #12
    "red",          #13
    "white",        #14
    "yellow"        #15
)

$Script:Back_Viper = 0   #0-Skip, 1-Default, 2-Safe, 3-Tweaked
$Script:Automated = $false
# Back Viper's Website
# http://www.blackviper.com/service-configurations/black-vipers-windows-10-service-configurations/

$WinEdition = gwmi win32_operatingsystem | % caption
#Pro = Microsoft Windows 10 Pro
#Home = Microsoft Windows 10 Home 

$BuildVer = [environment]::OSVersion.Version.build
# 14393 = anniversary update
# 10586 = first major update
# 10240 = first release

$ServicesList = @(
# Layout of Array (Numbers in = Service Type)
#(Service Name, Def-Home, Def-Pro , Safe, Tweaked)
('AJRouter',2,2,2,1),                                #AllJoyn Router Service
('ALG',2,2,2,1),                                     #Application Layer Gateway Service
('BthHFSrv',2,2,2,1),                                #Bluetooth Handsfree Service
('bthserv',2,2,2,1),                                 #Bluetooth Support Service
('PeerDistSvc',0,2,1,1),                             #BranchCache
('CertPropSvc',2,2,2,1),                             #Certificate Propagation
('NfsClnt',0,0,1,1),                                 #Client for NFS
('CDPUserSvc_',3,3,2,2),                             #Connected Devices Platform User Service_
('PimIndexMaintenanceSvc_',2,2,1,1),                 #Contact Data_
('TrkWks',3,3,3,1),                                  #Distributed Link Tracking Client
('dmwappushsvc',2,2,1,1),                            #Diagnostics Tracking Service (DiagTrack) or dmwappushsvc
('MapsBroker',4,4,1,1),                              #Downloaded Maps Manager
('DusmSvc',3,3,3,1),                                 #DusmSvc
('Fax',2,2,1,1),                                     #Fax
('lfsvc',2,2,1,1),                                   #Geolocation Service
('hkmsvc',0,2,2,1),                                  #Health Key and Certificate Management Removed in Creators Update
('HvHost',2,2,1,1),                                  #HV Host Service
('vmickvpexchange',2,2,1,1),                         #Hyper-V Data Exchange Service
('vmicguestinterface',2,2,1,1),                      #Hyper-V Guest Service Interface
('vmicshutdown',2,2,1,1),                            #Hyper-V Guest Shutdown Service
('vmicheartbeat',2,2,1,1),                           #Hyper-V Heartbeat Service
('vmicvmsession',2,2,1,1),                           #Hyper-V PowerShell Direct Service
('vmicrdv',2,2,1,1),                                 #Hyper-V Remote Desktop Virtualization Service
('vmictimesync',2,2,1,1),                            #Hyper-V Time Synchronization Service
('vmicvss',2,2,1,1),                                 #Hyper-V Volume Shadow Copy Requestor
('irmon',2,2,1,1),                                   #Infrared monitor service
('SharedAccess',2,2,1,1),                            #Internet Connection Sharing (ICS)
('iphlpsvc',3,3,3,1),                                #IP Helper
('IpxlatCfgSvc',2,2,2,1),                            #IP Translation Configuration Service
('wlpasvc',2,2,2,1),                                 #LPA Service
('diagnosticshub.standardcollector.service',2,2,1,1),#Microsoft (R) Diagnostics Hub Standard Collector Service
('MSiSCSI',2,2,1,1),                                 #Microsoft iSCSI Initiator Service
('SmsRouter',2,2,1,1),                               #Microsoft Windows SMS Router Service
('NaturalAuthentication',2,2,2,1),                   #Natural Authentication
('NetTcpPortSharing',2,2,1,1),                       #Net.Tcp Port Sharing Service
('Netlogon',2,2,2,1),                                #Netlogon
('NcdAutoSetup',2,2,2,1),                            #Network Connected Devices Auto-Setup
('NcbService',2,2,2,1),                              #Network Connection Broker
('CscService',0,2,1,1),                              #Offline Files
('SEMgrSvc',2,2,1,1),                                #Payments and NFC/SE Manager
('PhoneSvc',2,2,1,1),                                #Phone Service
('RmSvc',2,2,1,1),                                   #Radio Management Service
('SessionEnv',2,2,2,1),                              #Remote Desktop Configuration
('TermService',2,2,2,1),                             #Remote Desktop Services
('UmRdpService',2,2,2,1),                            #Remote Desktop Services UserMode Port Redirector
('RpcLocator',2,2,1,1),                              #Remote Procedure Call (RPC) Locator
('RetailDemo',2,2,1,1),                              #Retail Demo Service
('SensorDataService',2,2,1,1),                       #Sensor Data Service
('SensrSvc',2,2,1,1),                                #Sensor Monitoring Service
('SensorService',2,2,1,1),                           #Sensor Service
('ScDeviceEnum',2,2,1,1),                            #Smart Card Device Enumeration Service
('SCPolicySvc',2,2,1,1),                             #Smart Card Removal Policy
('SNMPTRAP',2,2,1,1),                                #SNMP Trap
('StorSvc',2,2,2,1),                                 #Storage Service
('TabletInputService',2,2,1,1),                      #Touch Keyboard and Handwriting Panel Service
('WebClient',2,2,2,1),                               #WebClient
('WFDSConSvc',2,2,1,1),                              #WFDSConMgrSvc
('WbioSrvc',2,2,2,1),                                #Windows Biometric Service
('FrameServer',2,2,1,1),                             #Windows Camera Frame Server
('wcncsvc',2,2,2,1),                                 #Windows Connect Now - Config Registrar
('wisvc',2,2,1,1),                                   #Windows Insider Service
('WMPNetworkSvc',2,2,1,1),                           #Windows Media Player Network Sharing Service
('icssvc',2,2,1,1),                                  #Windows Mobile Hotspot Service
('Wms',0,3,1,1),                                     #Windows MultiPoint Server Host Service
('WmsRepair',0,3,1,1),                               #Windows MultiPoint Server Repair Service
('WinRM',2,2,1,1),                                   #Windows Remote Management (WS-Management)
('spectrum',2,2,2,1),                                #Windows Spectrum
('workfolderssvc',2,2,1,1),                          #Work Folders
('WwanSvc',2,2,1,1),                                 #WWAN AutoConfig
('XblAuthManager',2,2,1,1),                          #Xbox Live Auth Manager
('XblGameSave',2,2,1,1),                             #Xbox Live Game Save
('XboxNetApiSvc',2,2,1,1))                           #XboxNetApiSvc

$ServiceLen = $ServicesList.length

$ServicesTypeList = @(
    '',          #0 -None (Not Installed, Default Only)
    'disabled',  #1 -Disable
    'manual',    #2 -Manual
    'automatic', #3 -Auto Normal
    'automatic'  #4 -Auto Delay
)

Function ServiceSet([Int]$ServiceVal){
    Clear-Host
    Write-Host "Changing Service Please wait..."
    for ($i=0; $i -ne $ServiceLen; $i++) {
        $ServiceT = $ServicesList[$i][$ServiceVal]
        $ServiceName = $ServicesList[$i][0]
        $ServiceNameFull = GetServiceNameFull $ServiceName
        $ServiceType = $ServicesTypeList[$ServiceT]
        $ServiceCurrType = (Get-Service $ServiceNameFull).StartType
        If ((ServiceCheck $ServiceNameFull $ServiceType $ServiceCurrType) -eq $True){
            If($ServiceT -In 1..3){
                Write-Host $ServiceNameFull "-" $ServiceCurrType "->" $ServiceType
                Set-Service $ServiceNameFull -StartupType $ServiceType
            } ElseIf($ServiceT -eq 4){
                Write-Host $ServiceNameFull "-" $ServiceCurrType "->" $ServiceType" (Delayed Start)"
                Set-Service $ServiceNameFull -StartupType $ServiceType
                $RegPath = "HKLM\System\CurrentControlSet\Services\"+($ServiceNameFull)
                Set-ItemProperty -Path $RegPath -Name "DelayedAutostart" -Type DWORD -Value 1
            }
        } Else {
            Write-Host $ServiceNameFull "is already" $ServiceCurrType
        }
    }
    Write-Host "Service Changed..."
    If($Automated -ne $true){
        Write-Host "Press any key to close..."
        $key = $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown,AllowCtrlC")
    }
    Exit
}

Function GetServiceNameFull([String]$ServiceN){
    $ServiceR = $ServiceN
    If($ServiceN -eq 'CDPUserSvc_'){
        $ServiceR = Get-Service | Where-Object {$_.Name -like "CDPUserSvc_*"}
    } ElseIf($ServiceN -eq 'PimIndexMaintenanceSvc_'){
        $ServiceR = Get-Service | Where-Object {$_.Name -like "PimIndexMaintenanceSvc_*"}
    }
    Return $ServiceR
}

Function ServiceCheck([string] $S_Name, [string]$S_Type, [string]$C_Type) {
    [bool] $ReturnV = $False
    If (Get-WmiObject -Class Win32_Service -Filter "Name='$S_Name'" ) {
        If($S_Type -ne $C_Type){
            $ReturnV = $True
            If($S_Name -eq 'lfsvc'){
                #Has to be removed or cant change service 
                # from disabled to anything else (Known Bug)
                Remove-Item -Path "HKLM:\SYSTEM\CurrentControlSet\Services\lfsvc\TriggerInfo\3"  -recurse  -Force
            }
        }
    }
    Return $ReturnV
}

function Black_Viper_Set ([Int]$Back_Viper){
    If ($Back_Viper -eq 1){
        If($WinEdition -eq "Microsoft Windows 10 Home"){
            ServiceSet $Back_Viper
        } ElseIf($WinEdition -eq "Microsoft Windows 10 Pro"){
            ServiceSet ($Back_Viper+1) 
        }
    } ElseIf ($Back_Viper -In 2..3){
        ServiceSet ($Back_Viper+1)
    }
}

If($WinEdition -eq "Microsoft Windows 10 Home" -or $WinEdition -eq "Microsoft Windows 10 Pro"){
    If ($SettingImp -ne $null -and $SettingImp){
        $Automated = $true
        If($SettingImp -In 1..3){
            Black_Viper_Set $SettingImp
        } ElseIf($SettingImp.ToLower() -eq "default"){
            Black_Viper_Set 1
        } ElseIf($SettingImp.ToLower() -eq "safe"){
            Black_Viper_Set 2
        } ElseIf($SettingImp.ToLower() -eq "tweaked"){
            Black_Viper_Set 3
        } Else{
            Write-Host "Invalid Selection" -ForegroundColor Blue -BackgroundColor Black
            Write-Host ""
            Write-Host "Valid Selections are:" -ForegroundColor Blue -BackgroundColor Black
            Write-Host "1 or Default" -ForegroundColor Green -BackgroundColor Black
            Write-Host "2 or Safe" -ForegroundColor Yellow -BackgroundColor Black
            Write-Host "3 or Tweaked" -ForegroundColor Red -BackgroundColor Black
            Write-Host ""
            Write-Host "Press Any key to Close..." -ForegroundColor White -BackgroundColor Black
            $key = $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown,AllowCtrlC")
        }
    } Else{
        TOS
    }
} Else {
    Write-Host "Websites:"
    Write-Host "https://github.com/madbomb122/"
    Write-Host "http://www.blackviper.com/"
    Write-Host ""
    Write-Host "Not a Valid OS for this Script." -ForegroundColor Red -BackgroundColor Black
    Write-Host "Win 10 Home and Pro Only"
    Write-Host ""
    Write-Host "Press Any key to Close..." -ForegroundColor White -BackgroundColor Black
    $key = $host.UI.RawUI.ReadKey("NoEcho,IncludeKeyDown,AllowCtrlC")
}
